src/main/java/io/vertx/core/impl/FutureImpl.java

```diff
@@ -78,29 +80,48 @@ public synchronized boolean isComplete() {
     return failed || succeeded;
   }
 
-  /**
-   * Set a handler for the result. It will get called when it's complete
-   */
-  public Future<T> setHandler(Handler<AsyncResult<T>> handler) {
-    boolean callHandler;
+  @Override
+  public Future<T> onComplete(Handler<AsyncResult<T>> h) {
+    if (h == null) {
+      throw new NullPointerException("No null handler accepted");
+    }
     synchronized (this) {
-      if (handler == null) {
-        this.handler = null;
-        return this;
-      } else {
-        callHandler = isComplete();
-        if (!callHandler) {
-          this.handler = handler;
+      if (!isComplete()) {
+        if (handler == null) {
+          handler = h;
+        } else {
+          addHandler(h);
         }
+        return this;
       }
     }
-    if (callHandler) {
-      dispatch(handler);
-    }
+    dispatch(h);
```

Comment:
Isn't there a chance for a concurrent modification exception when iterating through the handlers? The list addition is in a `synchronized` block, but all for loops may throw a `java.util.ConcurrentModificationException` if someone added a handler in-between.

I reckon this is highly unlikely to happen, but I'd like to have another opinion 😉

Reply:
I think it is not possible because the `tryComplete` / `tryFail` method will set the `handler` field to `null` in the synchronized block. Further calls to `onComplete` after will call `dispatch` and not create or set the `handler` field using the same synchronized monitor.

Reply:
Any further comment on this ?

Reply:
Nope